### DISCUSSION TITLE
Perspective-python installed successfully but won't import in a Jupyter notebook due to a binding issue

### DISCUSSION THREAD

**Discussion by bili-tufli** [perspective-dev/perspective#1360] [Q&A] [answered]
## import perspective --> CRITICAL:root: Failed to import C++ bindings for Perspective

Please advise how to fix " Failed to import C++ bindings for Perspective" when importing perspective into a jupyter notebook. 

After I installed perspective, I also installed the two required widgets listed below, which required an install of `Node.JS (global), jupyter, jupyterlab, and npm` in my virtual environment. Is it possible that these installs might have negatively affected the local environment in which I had installed perspective?
```
jupyter labextension install @finos/perspective-jupyterlab
jupyter labextension install @jupyter-widgets/jupyterlab-manager
```

`(env_perspective) C:\Users\joedoe\Documents\GitHub\project_perspective>pip check
No broken requirements found.`

(env_perspective) C:\Users\jowdoe\Documents\GitHub\project_perspective>perspective-python
'perspective-python' is not recognized as an internal or external command,
operable program or batch file.

Python 3.7.9
jupyter core     : 4.7.1
jupyter-notebook : 6.2.0
qtconsole        : 5.0.2
ipython          : 7.20.0
ipykernel        : 5.4.3
jupyter client   : 6.1.11
jupyter lab      : 3.0.8
nbconvert        : 6.0.7
ipywidgets       : 7.6.3
nbformat         : 5.1.2
traitlets        : 5.0.5
jupyter labextension 3.0.8

## ERROR on  `import perspective`

``` CRITICAL:root:Failed to import C++ bindings for Perspective probably as it could not be built for your architecture (check install logs for more details).
Traceback (most recent call last):
  File "C:\Users\joedoe\Documents\GitHub\project_perspective\env_perspective\lib\site-packages\perspective\libpsp.py", line 21, in <module>
    from .table import *  # noqa: F401, F403
  File "C:\Users\joedoe\Documents\GitHub\project_perspective\env_perspective\lib\site-packages\perspective\table\__init__.py", line 9, in <module>
    from .table import Table
  File "C:\Users\joedoe\Documents\GitHub\project_perspective\env_perspective\lib\site-packages\perspective\table\table.py", line 11, in <module>
    from .view import View
  File "C:\Users\joedoe\Documents\GitHub\project_perspective\env_perspective\lib\site-packages\perspective\table\view.py", line 15, in <module>
    from ._data_formatter import to_format, _parse_format_options
  File "C:\Users\joedoe\Documents\GitHub\project_perspective\env_perspective\lib\site-packages\perspective\table\_data_formatter.py", line 12, in <module>
    from .libbinding import (
ModuleNotFoundError: No module named 'perspective.table.libbinding'
CRITICAL:root:You can still use `PerspectiveWidget` in client mode using JupyterLab.
```

**Comment by bili-tufli** [comment]
pip list 
`Package             Version
------------------- ---------
anyio               2.1.0
argon2-cffi         20.1.0
async-generator     1.10
attrs               20.3.0
Babel               2.9.0
backcall            0.2.0
bleach              3.3.0
certifi             2020.12.5
cffi                1.14.5
chardet             4.0.0
colorama            0.4.4
decorator           4.4.2
defusedxml          0.6.0
entrypoints         0.3
future              0.18.2
idna                2.10
importlib-metadata  3.4.0
ipykernel           5.4.3
ipython             7.20.0
ipython-genutils    0.2.0
ipywidgets          7.6.3
jedi                0.18.0
Jinja2              2.11.3
json5               0.9.5
jsonschema          3.2.0
jupyter             1.0.0
jupyter-client      6.1.11
jupyter-console     6.2.0
jupyter-core        4.7.1
jupyter-server      1.4.0
jupyterlab          3.0.8
jupyterlab-pygments 0.1.2
jupyterlab-server   2.3.0
jupyterlab-widgets  1.0.0
MarkupSafe          1.1.1
mistune             0.8.4
nbclassic           0.2.6
nbclient            0.5.2
nbconvert           6.0.7
nbformat            5.1.2
nest-asyncio        1.5.1
nodejs              0.1.1
notebook            6.2.0
npm                 0.1.1
numpy               1.20.1
optional-django     0.1.0
packaging           20.9
pandas              1.2.2
pandocfilters       1.4.3
parso               0.8.1
perspective-python  0.6.0
pickleshare         0.7.5
pip                 20.1.1
prometheus-client   0.9.0
prompt-toolkit      3.0.16
pyarrow             0.16.0
pybind11            2.6.2
pybind11-global     2.6.2
pycparser           2.20
Pygments            2.8.0
pyparsing           2.4.7
pyrsistent          0.17.3
python-dateutil     2.8.1
pytz                2021.1
pywin32             300
pywinpty            0.5.7
pyzmq               22.0.3
qtconsole           5.0.2
QtPy                1.9.0
requests            2.25.1
Send2Trash          1.5.0
setuptools          53.0.0
six                 1.15.0
sniffio             1.2.0
terminado           0.9.2
testpath            0.4.4
tornado             6.1
traitlets           5.0.5
typing-extensions   3.7.4.3
urllib3             1.26.3
wcwidth             0.2.5
webencodings        0.5.1
widgetsnbextension  3.5.1
zipp                3.4.0`

**Comment by bili-tufli** [comment]
(env_perspective) C:\Users\joedoe\Documents\GitHub\project_perspective>pip install --upgrade perspective-python
Requirement already up-to-date: perspective-python in ...\site-packages (0.6.0)
Requirement already satisfied, skipping upgrade: pyarrow<1,>=0.16.0 in ...\site-packages (from perspective-python) (0.16.0)
Requirement already satisfied, skipping upgrade: future>=0.16.0 in ...\site-packages (from perspective-python) (0.18.2)
Requirement already satisfied, skipping upgrade: numpy>=1.13.1 in ...\site-packages (from perspective-python) (1.20.1)
Requirement already satisfied, skipping upgrade: ipywidgets>=7.5.1 in ...\site-packages (from perspective-python) (7.6.3)
Requirement already satisfied, skipping upgrade: six>=1.11.0 in ...\site-packages (from perspective-python) (1.15.0)
Requirement already satisfied, skipping upgrade: traitlets>=4.3.2 in ...\site-packages (from perspective-python) (5.0.5)
Requirement already satisfied, skipping upgrade: pandas>=0.22.0 in ...\site-packages (from perspective-python) (1.2.2)
Requirement already satisfied, skipping upgrade: tornado>=4.5.3 in ...\site-packages (from perspective-python) (6.1)
Requirement already satisfied, skipping upgrade: python-dateutil>=2.8.0 in ...\site-packages (from perspective-python) (2.8.1)
Requirement already satisfied, skipping upgrade: widgetsnbextension~=3.5.0 in ...\site-packages (from ipywidgets>=7.5.1->perspective-python) (3.5.1)
Requirement already satisfied, skipping upgrade: ipykernel>=4.5.1 in ...\site-packages (from ipywidgets>=7.5.1->perspective-python) (5.4.3)
Requirement already satisfied, skipping upgrade: jupyterlab-widgets>=1.0.0; python_version >= "3.6" in ...\site-packages (from ipywidgets>=7.5.1->perspective-python) (1.0.0)
Requirement already satisfied, skipping upgrade: ipython>=4.0.0; python_version >= "3.3" in ...\site-packages (from ipywidgets>=7.5.1->perspective-python) (7.20.0)
Requirement already satisfied, skipping upgrade: nbformat>=4.2.0 in ...\site-packages (from ipywidgets>=7.5.1->perspective-python) (5.1.2)
Requirement already satisfied, skipping upgrade: ipython-genutils in ...\site-packages (from traitlets>=4.3.2->perspective-python) (0.2.0)
Requirement already satisfied, skipping upgrade: pytz>=2017.3 in ...\site-packages (from pandas>=0.22.0->perspective-python) (2021.1)
Requirement already satisfied, skipping upgrade: notebook>=4.4.1 in ...\site-packages (from widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (6.2.0)
Requirement already satisfied, skipping upgrade: jupyter-client in ...\site-packages (from ipykernel>=4.5.1->ipywidgets>=7.5.1->perspective-python) (6.1.11)
Requirement already satisfied, skipping upgrade: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in ...\site-packages (from ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (3.0.16)
Requirement already satisfied, skipping upgrade: colorama; sys_platform == "win32" in ...\site-packages (from ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (0.4.4)
Requirement already satisfied, skipping upgrade: jedi>=0.16 in ...\site-packages (from ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (0.18.0)
Requirement already satisfied, skipping upgrade: backcall in ...\site-packages (from ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (0.2.0)
Requirement already satisfied, skipping upgrade: setuptools>=18.5 in ...\site-packages (from ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (53.0.0)
Requirement already satisfied, skipping upgrade: pickleshare in ...\site-packages (from ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (0.7.5)
Requirement already satisfied, skipping upgrade: decorator in ...\site-packages (from ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (4.4.2)
Requirement already satisfied, skipping upgrade: pygments in ...\site-packages (from ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (2.8.0)
Requirement already satisfied, skipping upgrade: jupyter-core in ...\site-packages (from nbformat>=4.2.0->ipywidgets>=7.5.1->perspective-python) (4.7.1)
Requirement already satisfied, skipping upgrade: jsonschema!=2.5.0,>=2.4 in ...\site-packages (from nbformat>=4.2.0->ipywidgets>=7.5.1->perspective-python) (3.2.0)
Requirement already satisfied, skipping upgrade: nbconvert in ...\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (6.0.7)
Requirement already satisfied, skipping upgrade: Send2Trash>=1.5.0 in ...\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (1.5.0)
Requirement already satisfied, skipping upgrade: jinja2 in ...\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (2.11.3)
Requirement already satisfied, skipping upgrade: pyzmq>=17 in ...\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (22.0.3)
Requirement already satisfied, skipping upgrade: argon2-cffi in ...\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (20.1.0)
Requirement already satisfied, skipping upgrade: prometheus-client in ...\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.9.0)
Requirement already satisfied, skipping upgrade: terminado>=0.8.3 in ...\site-packages (from notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.9.2)
Requirement already satisfied, skipping upgrade: wcwidth in ...\site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (0.2.5)
Requirement already satisfied, skipping upgrade: parso<0.9.0,>=0.8.0 in ...\site-packages (from jedi>=0.16->ipython>=4.0.0; python_version >= "3.3"->ipywidgets>=7.5.1->perspective-python) (0.8.1)
Requirement already satisfied, skipping upgrade: pywin32>=1.0; sys_platform == "win32" in ...\site-packages (from jupyter-core->nbformat>=4.2.0->ipywidgets>=7.5.1->perspective-python) (300)
Requirement already satisfied, skipping upgrade: importlib-metadata; python_version < "3.8" in ...\site-packages (from jsonschema!=2.5.0,>=2.4->nbformat>=4.2.0->ipywidgets>=7.5.1->perspective-python) (3.4.0)
Requirement already satisfied, skipping upgrade: attrs>=17.4.0 in ...\site-packages (from jsonschema!=2.5.0,>=2.4->nbformat>=4.2.0->ipywidgets>=7.5.1->perspective-python) (20.3.0)
Requirement already satisfied, skipping upgrade: pyrsistent>=0.14.0 in ...\site-packages (from jsonschema!=2.5.0,>=2.4->nbformat>=4.2.0->ipywidgets>=7.5.1->perspective-python) (0.17.3)
Requirement already satisfied, skipping upgrade: testpath in ...\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.4.4)
Requirement already satisfied, skipping upgrade: bleach in ...\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (3.3.0)
Requirement already satisfied, skipping upgrade: defusedxml in ...\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.6.0)
Requirement already satisfied, skipping upgrade: entrypoints>=0.2.2 in ...\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.3)
Requirement already satisfied, skipping upgrade: mistune<2,>=0.8.1 in ...\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.8.4)
Requirement already satisfied, skipping upgrade: jupyterlab-pygments in ...\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.1.2)
Requirement already satisfied, skipping upgrade: pandocfilters>=1.4.1 in ...\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (1.4.3)
Requirement already satisfied, skipping upgrade: nbclient<0.6.0,>=0.5.0 in ...\site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.5.2)
Requirement already satisfied, skipping upgrade: MarkupSafe>=0.23 in ...\site-packages (from jinja2->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (1.1.1)
Requirement already satisfied, skipping upgrade: cffi>=1.0.0 in ...\site-packages (from argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (1.14.5)
Requirement already satisfied, skipping upgrade: pywinpty>=0.5; os_name == "nt" in ...\site-packages (from terminado>=0.8.3->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.5.7)
Requirement already satisfied, skipping upgrade: typing-extensions>=3.6.4; python_version < "3.8" in ...\site-packages (from importlib-metadata; python_version < "3.8"->jsonschema!=2.5.0,>=2.4->nbformat>=4.2.0->ipywidgets>=7.5.1->perspective-python) (3.7.4.3)
Requirement already satisfied, skipping upgrade: zipp>=0.5 in ...\site-packages (from importlib-metadata; python_version < "3.8"->jsonschema!=2.5.0,>=2.4->nbformat>=4.2.0->ipywidgets>=7.5.1->perspective-python) (3.4.0)
Requirement already satisfied, skipping upgrade: webencodings in ...\site-packages (from bleach->nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (0.5.1)
Requirement already satisfied, skipping upgrade: packaging in ...\site-packages (from bleach->nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (20.9)
Requirement already satisfied, skipping upgrade: async-generator in ...\site-packages (from nbclient<0.6.0,>=0.5.0->nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (1.10)
Requirement already satisfied, skipping upgrade: nest-asyncio in ...\site-packages (from nbclient<0.6.0,>=0.5.0->nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (1.5.1)
Requirement already satisfied, skipping upgrade: pycparser in ...\site-packages (from cffi>=1.0.0->argon2-cffi->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (2.20)
Requirement already satisfied, skipping upgrade: pyparsing>=2.0.2 in ...\site-packages (from packaging->bleach->nbconvert->notebook>=4.4.1->widgetsnbextension~=3.5.0->ipywidgets>=7.5.1->perspective-python) (2.4.7)

**Comment by bili-tufli** [comment]
Unsuccessfully tried the following solution #[1108](https://github.com/finos/perspective/issues/1108#):
`pip install --no-clean -v perspective-python`

After re-installing this way, however, I cannot find any .dll files to copy/move to the 'table' folder. 
Then, tried uninstalling and re-installing perspective-python. No luck - still cannot import perspective in jupyter in a virtual environment (although installed successfully).

**Comment by timkpaine** [comment]
The logs above indicate no reinstall occurred, you probably need --force-reinstall

**Comment by bili-tufli** [comment]
> The logs above indicate no reinstall occurred, you probably need --force-reinstall

Thank you, @timkpaine , for your prompt response. 
Actually, I tried pip install --upgrade --force-reinstall --no clean --verbose perspective-python
Unfortunately, that did not solve the problem.

Then I tried pip uninstall perspective-python followed by pip install --no clean --verbose perspective-python
Still no luck....

Is there anything else I can try?

**Comment by bili-tufli** [comment]
After reading about issues installing perspective in a virtual environment, I tried installing it in a global environment. Still no  luck...

**Comment by timkpaine** [comment, marked as answer]
try building from source. without logs showing a build time problem theres not much we can do to help. once you've built and installed, ensure that all the dependencies are in place including dlls (these need to be on your path on windows).
